src/goto-instrument/contracts/assigns.cpp

```diff
@@ -132,7 +132,8 @@ void assigns_clauset::remove_from_local_write_set(const exprt &expr)
   local_write_set.erase(targett(*this, expr));
 }
 
-goto_programt assigns_clauset::generate_havoc_code() const
+goto_programt
+assigns_clauset::generate_havoc_code(const source_locationt &location) const
 {
   modifiest modifies;
   for(const auto &target : global_write_set)
```

Comment:
I thought we didn't need separate local and global write sets now. Is that a future PR. Also, I see the addresses in the code, don't we need more than just the address to know what to havoc (location vs object containing location)?

Reply:
> I thought we didn't need separate local and global write sets now. Is that a future PR.

We now only track one `write_set` now. I was implementing this change on a separate PR, but I pulled the commit to this one.

> Also, I see the addresses in the code, don't we need more than just the address to know what to havoc (location vs object containing location)?

The correct answer here is that a proper implementation of CARs will give us all necessary info for enforcement and replacement. Now, we keep track of the address in a `exprt` which will also give us the source location and the object associated with this address (we use this info to havoc).